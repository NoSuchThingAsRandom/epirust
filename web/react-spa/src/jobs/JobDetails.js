/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import TimeSeries from "../time-series";
import "./job-details.scss"
import {NavItem} from "../common/NavItem";
import {Route, Switch} from "react-router-dom";
import GridPage from "../grid";

export const JobDetails = ({simulationId}) => {
  const linkPrefix = `/jobs/${simulationId}`;

  return (
    <div className="job-details" style={{paddingTop:4}}>
      <h4>Simulation: {simulationId}</h4>
      <ul className="nav nav-tabs">
        <NavItem name="Time Series" linksTo={`${linkPrefix}/time-series`} activeOnExactMatch={true}/>
        <NavItem name="Grid" linksTo={`${linkPrefix}/grid`}/>
        <NavItem name="Config" linksTo={`${linkPrefix}/config`}/>
      </ul>
      <Switch>
        <Route exact path={"/jobs/:id/time-series"}><TimeSeries simulationId={simulationId}/></Route>
        <Route exact path={"/jobs/:id/grid"}><GridPage/></Route>
        <Route exact path={"/jobs/:id/config"}>Config</Route>
      </Switch>
    </div>
  )
};
