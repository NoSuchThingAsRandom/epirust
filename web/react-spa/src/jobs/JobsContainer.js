/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {Redirect} from "react-router-dom";
import {Jobs} from "./Jobs";
import React from "react";

export const JobsContainer = ({activeJobId, currentView, jobs}) => {
  if (!activeJobId)
    return (<Redirect to={`/jobs/${jobs[0]._id}/time-series`}/>);
  if (!currentView)
    return (<Redirect to={`/jobs/${activeJobId}/time-series`}/>);

  const activeJob = jobs.find(j => j._id === activeJobId)
  if (!activeJob) return <div>Invalid job requested</div>
  else return <Jobs jobs={jobs} activeJob={activeJob}/>
}
