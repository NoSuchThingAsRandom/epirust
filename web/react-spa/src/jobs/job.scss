/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

.borderless {
  border: none;
}

.card-body-less-padding {
  padding: 0.5rem;
}

.item-less-padding {
  padding: 0.25rem 1.25rem
}

.job-status {
  font-size: 22px;
  font-weight: bold;
  line-height: 26px;
}

.minor-details {
  font-size: 16px;
}

.no-link-formatting {
  color: white;
  text-decoration: none;
  &:focus, &:hover, &:visited, &:link, &:active {
    text-decoration: none;
    color: white;
  }
}