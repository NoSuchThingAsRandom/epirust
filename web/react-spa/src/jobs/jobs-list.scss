/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

.jobs-list {
  .tabs {
  }
  .details {
    padding: 0 20px;
  }
}

.scrollable {
  height: 90vh;
  position: relative;
  overflow: auto;
}

.running {
  background-color: #F6BB43;
}

.failed {
  background-color: #F9694C;
}

.finished {
  background-color: #8BC253;
}

.in-queue {
  background-color: #3CAEDA;
}

.left-border {
  border-style: none none none solid;
  border-color: black;
}

.padded-list {
  padding: 0 16px 0 0;
}