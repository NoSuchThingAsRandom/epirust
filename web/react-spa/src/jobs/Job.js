/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import './job.scss'
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

export const Job = ({ simulationId, status, isActive = false }) => {
  const statusText = {
    "running": "In-Progress",
    "finished": "Finished",
    "failed": "Failed",
    "in-queue": "In-Queue"
  };

  return (
    <li className="list-group-item borderless item-less-padding">

      <Link to={`/jobs/${simulationId}`} className={'no-link-formatting'}>
        <div className={`simulation-tab ${isActive ? "active shadow" : ""} ${status}`}>
          <div className="card-body card-body-less-padding">
            <p data-testid={`job-status-${simulationId}`} className="job-status">{statusText[status]}</p>
            <p className="minor-details">{`Id: `}<code>{simulationId}</code></p>
          </div>
        </div>
      </Link>

    </li>
  );
};

Job.propTypes = {
  simulationId: PropTypes.number.isRequired,
  status: PropTypes.oneOf(["running", "finished", "failed", "in-queue"]).isRequired
};