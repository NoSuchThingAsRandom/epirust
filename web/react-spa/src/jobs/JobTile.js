/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import '../styles/components/jobs/job.scss'
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

export const JobTile = ({jobId, status, isActive }) => {
  const statusText = {
    "running": "In-Progress",
    "finished": "Finished",
    "failed": "Failed",
    "in-queue": "In-Queue"
  };

  return (
    <li className="list-group-item borderless item-less-padding">

      <Link to={`/jobs/${jobId}`} className={'no-link-formatting'}>
        <div className={`simulation-tab ${isActive ? "active shadow" : ""} ${status}`}>
          <div className="card-body card-body-less-padding">
            <p data-testid={`job-status-${jobId}`} className="job-status">{statusText[status]}</p>
            <p className="minor-details">{`Id: `}<code>{jobId}</code></p>
          </div>
        </div>
      </Link>

    </li>
  );
};

JobTile.propTypes = {
  jobId: PropTypes.string.isRequired,
  status: PropTypes.oneOf(["running", "finished", "failed", "in-queue"]).isRequired
};