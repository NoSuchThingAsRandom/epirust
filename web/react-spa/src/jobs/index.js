/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from "react";
import '../styles/components/jobs/jobs-list.scss'
import {useParams} from 'react-router-dom';
import {get} from "../common/apiCall";
import {reduceStatus} from "./JobTransformer";
import {LOADING_STATES} from "../common/constants";
import {JobsContainer} from "./JobsContainer";
import {LoadingComponent} from "../common/LoadingComponent";

export const JobsView = () => {
  const [loadingState, updateLoadingState] = useState(LOADING_STATES.LOADING)
  const [jobs, updateJobs] = useState([])
  const {id: activeJobId, view} = useParams();

  const refreshJobs = (jobsToProcess) => {
    const jobsToFetch = jobsToProcess.filter(j => j.status !== "finished")
    if (jobsToFetch.length > 0) {
      const url = `/jobs?jobIds=${jobsToFetch.map(j => j._id).join(",")}`
      get(url)
        .then(res => res.json())
        .then(jobsToUpdate => {
          updateJobs(prevJobs => {
            const updatedJobs = prevJobs.map(pj => {
              const jobToUpdate = jobsToUpdate.find(ju => ju._id === pj._id);
              return jobToUpdate ? reduceStatus(jobToUpdate) : pj;
            });

            setTimeout(() => refreshJobs(updatedJobs), 15000);
            return updatedJobs;
          })
        })
    }
  }

  useEffect(() => {
    get("/jobs")
      .then(res => res.json())
      .then(receivedJobs => {
        const convertedJobs = receivedJobs.map(reduceStatus).reverse();
        updateJobs(convertedJobs)
        updateLoadingState(LOADING_STATES.FINISHED)
        setTimeout(() => refreshJobs(convertedJobs), 15000)
      })
      .catch(() => updateLoadingState(LOADING_STATES.FAILED))
  }, []);

  const children = <JobsContainer activeJobId={activeJobId} currentView={view} jobs={jobs}/>;
  return <LoadingComponent loadingState={loadingState} children={children}/>;
}
