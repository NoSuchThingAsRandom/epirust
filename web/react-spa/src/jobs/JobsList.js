/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from "react";
import './jobs-list.scss'
import {Job} from "./Job";
import {JobDetails} from "./JobDetails";
import {Redirect, useParams} from 'react-router-dom';

export const JobsList = () => {
  const {id, view} = useParams();
  const [simulations, updateSimulations] = useState([]);

  useEffect(() => {
    fetch("http://localhost:3000/simulation/",
      {
        method: 'GET',
        headers: {'Content-Type': 'application/json'}
      })
      .then(value => value.json())
      .then(value => updateSimulations(value.reverse()))
  }, []);

  if (id && !view) {
    return (<Redirect to={`/jobs/${id}/time-series`}/>);
  } else {
    return (<div className="row jobs-list">
      <div className="col-3">
        <ul className="list-group scrollable">
          {simulations.map(s => <Job key={s.simulation_id} simulationId={s.simulation_id} status={s.status}/>)}
        </ul>
      </div>
      <div className="col-9">
        {id && <JobDetails simulationId={id}/>}
      </div>
    </div>);
  }
};
