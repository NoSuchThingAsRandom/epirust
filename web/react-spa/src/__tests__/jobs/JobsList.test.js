/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {JobsList} from "../../jobs/JobsList";
import React from "react";
import {render} from "@testing-library/react";
import {MemoryRouter} from "react-router-dom";

describe('JobsList', () => {
  let sarsTag = {id: "12", name: "SARS"}, covidTag = {id: "11", name: "covid-19"};
  let sarsJobId = "12345", covidJobId = "12344", sarsCovidJob = "12346", noTagsJob = '123';

  const testJobs = [
    {_id: covidJobId, status: "finished", config: {tags: [covidTag]}},
    {_id: sarsJobId, status: "finished", config: {tags: [sarsTag]}},
    {_id: sarsCovidJob, status: "finished", config: {tags: [covidTag, sarsTag]}},
    {_id: noTagsJob, status: "finished", config: {tags: []}}
  ];

  it('should render jobs with active job if valid active job', () => {
    const {asFragment} = render(<MemoryRouter><JobsList jobs={testJobs} activeJob={testJobs[1]}/></MemoryRouter>);
    expect(asFragment()).toMatchSnapshot()
  });

  it('should render jobs list with no active active job if it is invalid', () => {
    const {asFragment} = render(<MemoryRouter><JobsList jobs={testJobs} activeJob={null}/></MemoryRouter>);
    expect(asFragment()).toMatchSnapshot()
  });
});
