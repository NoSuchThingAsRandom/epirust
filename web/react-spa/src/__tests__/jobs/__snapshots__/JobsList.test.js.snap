// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobsList should fetch simulations when id and view is not selected 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row jobs-list"
      >
        <div
          class="col-3"
        >
          <ul
            class="list-group scrollable"
          >
            <li
              class="list-group-item borderless item-less-padding"
            >
              <a
                class="no-link-formatting"
                href="/jobs/1234"
              >
                <div
                  class="card finished"
                >
                  <div
                    class="card-body card-body-less-padding"
                  >
                    <div
                      class="job-status"
                    >
                      Finished
                    </div>
                    <div
                      class="minor-details"
                    >
                      Simulation Id: 
                      1234
                       
                    </div>
                  </div>
                </div>
              </a>
            </li>
          </ul>
        </div>
        <div
          class="col-9"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row jobs-list"
    >
      <div
        class="col-3"
      >
        <ul
          class="list-group scrollable"
        >
          <li
            class="list-group-item borderless item-less-padding"
          >
            <a
              class="no-link-formatting"
              href="/jobs/1234"
            >
              <div
                class="card finished"
              >
                <div
                  class="card-body card-body-less-padding"
                >
                  <div
                    class="job-status"
                  >
                    Finished
                  </div>
                  <div
                    class="minor-details"
                  >
                    Simulation Id: 
                    1234
                     
                  </div>
                </div>
              </div>
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-9"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
